### PR TITLE
[WIP] [doc] performance/scalability revamp

### PR DESCRIPTION
**Pull request by stas00**
moved from https://github.com/huggingface/transformers/pull/15213 so that we get the doc generator working

XXX: The previous PR has comments/suggestions that need to be integrated here

-----------------

@lvwerra and I are working on a massive performance/scalability docs revamp:

So the rough plan is to make custom plans for each of the combinations `[inference|training] * [1 gpu|many gpus|cpu]` so that it's very easy for the user to follow the instructions that are specific to their needs.

So the proposed doc layout is:

* performance.mdx (main entry point)
* perf_infer.mdx
   - perf_infer_cpu.mdx
   - perf_infer_gpu_many.mdx
   - perf_infer_gpu_one.mdx
* perf_train.mdx
   - perf_train_gpu_many.mdx
   - perf_train_gpu_one.mdx
* scalability.mdx (rename from parallelism.mdx) (XXX: to do)

See the PR's changes for a rough layout of the content.

One big question is this: At the moment everything is pytorch-centric, as we don't have any info on tf/flax. Down the road we will either inject tf/flax-specific instructions into the current docs, or perhaps it'd be better to have dedicated docs for pt/tf/flax. It'd help a lot to decide ahead of time to avoid document renaming and potentially breaking links. If we plan to have these PT-specific perhaps let's embed `_pt` in the filenames?

@lvwerra

### CODE REVIEW

**Comment by stas00** [comment]
Thank you for taking a stab at redesign, @lvwerra 

It's very clear we have fundamentally conflicting visions of how the performance documentation should be laid out. So it'd be wasteful to spend energy to try to come up with something that resonates for both of us.

I don't think this project can have 2 masterminds, and since I'm currently busy with BigScience training and we don't want this to fall between the cracks, I propose that if this works for you, please take over and do it in the unimpeded way that sounds good to you, @lvwerra.

Once you have completed what you know and you want me to fill in some gaps please let me know and I will do so.

**Comment by lvwerra** [comment]
Hi @LysandreJik and @sgugger, 

I revamped the documentation a bit and reduced the scope from all the documents @stas00 lined out to just the `performance.mdx` and the training on single/multi-GPU so we can converge on those before adding the others, which also prevents a monster PR. For some reason the preview of the docs still does not work - if you have any idea, let me know. 

To guide you a bit in the review:
- `performance.mdx` should be the entry point and give an overview
- `perf_train_gpu_single.mdx` shows all the tricks to efficiently train large models on a single GPU (gradient accumulation/checkpointing, optimizers, DeepSpeed Zero, data loaders etc.)
- `perf_train_gpu_single.mdx` shows all the tricks to efficiently train large models on a single GPU (mostly parallelism strategies: data, tensor, pipeline parallelism)
- `perf_hardware.mdx` is where I put some tips and tricks about custom hardware setups

In subsequent PRs we can:
- add more TensorFlow examples
- add TPU training guide
- add inference side
- add guides for train/inference on specialized hardware (Optimum)

This is the plan for the full documentation:

perf_overview

Let me know what you think about style and content of the documentation. There are still a few rough edges but I think it is good for a first review.

**Comment by sgugger** [comment]
The preview can't work if you don't rebase on master to have the new doc structure (everything nested in an "en" subfolder)

**Comment by lvwerra** [comment]
Thank you @sgugger it worked!

File: docs/source/en/_toctree.yml [hunk 1]
```diff
@@ -81,6 +77,16 @@
     title: "How to add a model to 🤗 Transformers?"
   - local: add_new_pipeline
     title: "How to add a pipeline to 🤗 Transformers?"
+  - local: fast_tokenizers
+    title: "Using tokenizers from 🤗 Tokenizers"
```
Comment (sgugger) [inline docs/source/en/_toctree.yml:81]: This change does not seem to be linked to this PR.

File: docs/source/en/perf_hardware.mdx [hunk 2]
```diff
@@ -0,0 +1,150 @@
+<!---
+Copyright 2021 The HuggingFace Team. All rights reserved.
```
Comment (sgugger) [inline docs/source/en/perf_hardware.mdx:2]:
```suggestion
Copyright 2022 The HuggingFace Team. All rights reserved.
```

File: docs/source/en/perf_train_gpu_many.mdx [hunk 3]
```diff
@@ -1,48 +1,158 @@
-<!---
-Copyright 2021 The HuggingFace Team. All rights reserved.
+<!--Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_many.mdx:1]:
```suggestion
<!--Copyright 2022 The HuggingFace Team. All rights reserved.
```

File: docs/source/en/perf_train_gpu_many.mdx [hunk 4]
```diff
@@ -1,48 +1,158 @@
...
-4. Zero Redundancy Optimizer (ZeRO) - Also performs sharding of the tensors somewhat similar to TP, except the whole tensor gets reconstructed in time for a forward or backward computation, therefore the model doesn't need to be modified. It also supports various offloading techniques to compensate for limited GPU memory.
-5. Sharded DDP - is another name for the foundational ZeRO concept as used by various other implementations of ZeRO.
 
 
 ## Data Parallelism
 
 Most users with just 2 GPUs already enjoy the increased training speed up thanks to `DataParallel` (DP) and `DistributedDataParallel` (DDP) that are almost trivial to use. This is a built-in feature of Pytorch.
 
+### DP vs DDP
+
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_many.mdx:72]: This should begin by replicating the warning from PyTorch encouraging users to use DDP over DP. DDP is better maintained by PyTorch and works for all models. Some of the models will fail with DP however.

File: docs/source/en/perf_train_gpu_one.mdx [hunk 5]
```diff
@@ -0,0 +1,721 @@
+<!--Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_one.mdx:1]:
```suggestion
<!--Copyright 2022 The HuggingFace Team. All rights reserved.
```

File: docs/source/en/perf_train_gpu_one.mdx [hunk 6]
```diff
@@ -0,0 +1,721 @@
...
+6. functionality-specific memory
+
+A typical model trained in mixed precision with AdamW requires 18 bytes per model parameter plus activation memory. For inference there are no optimizer states and gradients, so we can subtract those. And thus we end up with 6 bytes per model parameter for mixed precision inference, plus activation memory.
+
+Let's look at the details.
+
+**Model Weights:**
+
+- 4 bytes * number of parameters for fp32 training
+- 6 bytes * number of parameters for mixed precision training
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_one.mdx:215]:
```suggestion
- 6 bytes * number of parameters for mixed precision training (maintains a model in fp32 and one in fp16 in memory)
```

File: docs/source/en/perf_train_gpu_one.mdx [hunk 7]
```diff
@@ -0,0 +1,721 @@
...
+
+**Optimizer States:**
+
+- 8 bytes * number of parameters for normal AdamW (maintains 2 states)
+- 2 bytes * number of parameters for 8-bit AdamW optimizers like [bitsandbytes](https://github.com/facebookresearch/bitsandbytes)
+- 4 bytes * number of parameters for optimizers like SGD (maintains only 1 state)
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_one.mdx:221]:
```suggestion
- 4 bytes * number of parameters for optimizers like SGD with momentum (maintains only 1 state)
```
Vanilla SGD does not have any state.

File: docs/source/en/perf_train_gpu_one.mdx [hunk 8]
```diff
@@ -0,0 +1,721 @@
...
+
+**Gradients**
+
+- 4 bytes * number of parameters for either fp32 or mixed precision training
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_one.mdx:225]:
```suggestion
- 4 bytes * number of parameters for either fp32 or mixed precision training (gradients are always kept in fp32)
```

File: docs/source/en/perf_train_gpu_one.mdx [hunk 9]
```diff
@@ -0,0 +1,721 @@
...
+The idea of mixed precision training is that no all variables need to be stored in full (32-bit) floating point precision. If we can reduce the precision the variales and their computations are faster. Here are the commonly used floating point data types choice of which impacts both memory usage and throughput:
+
+- fp32 (`float32`)
+- fp16 (`float16`)
+- bf16 (`bfloat16`)
+- tf32 (CUDA internal data type)
+
+Here is a diagram that shows how these data types correlate to each other.
+
+**TODO:** Can we make this graphic prettier maybe?
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_one.mdx:323]: Flagging this is here.

File: docs/source/en/perf_train_gpu_one.mdx [hunk 10]
```diff
@@ -0,0 +1,721 @@
...
+When this is done CUDA will automatically switch to using tf32 instead of fp32 where it's possible. This, of course, assumes that the used GPU is from the Ampere series.
+
+Like all cases with reduced precision this may or may not be satisfactory for your needs, so you have to experiment and see. According to [NVIDIA research](https://developer.nvidia.com/blog/accelerating-ai-training-with-tf32-tensor-cores/) the majority of machine learning training shouldn't be impacted and showed the same perplexity and convergence as the fp32 training.
+
+If you're already using fp16 or bf16 mixed precision it may help with the throughput as well.
+
+You can enable this mode in the 🤗 Trainer with: 
+    ```python
+    TrainingArguments(tf32=True)
+    ```
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_one.mdx:401]:
```suggestion
```python
TrainingArguments(tf32=True)
```
```

File: docs/source/en/perf_train_gpu_one.mdx [hunk 11]
```diff
@@ -0,0 +1,721 @@
...
+Samples/second: 10.35
+GPU memory occupied: 5363 MB.
+```
+
+Again, we get about a 3x memory improvement and even slightly higher throughput as using Adafactor. So we have seen how we can optimize the memory footprint of large models. The following plot summarizes all our experiments:
+
+![png](https://huggingface.co/datasets/lvwerra/repo-images/raw/main/gpu-memory-savings.png)
+
+### `_multi_tensor`
+pytorch-nightly introduced `torch.optim._multi_tensor` which should significantly speed up the optimizers for situations with lots of small feature tensors. It should eventually become the default, but if you want to experiment with it sooner and don't mind using the bleed-edge, see: https://github.com/huggingface/transformers/issues/9965
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_one.mdx:570]: This is one year old, is it really the bleeding edge?

Comment (stas00) [reply docs/source/en/perf_train_gpu_one.mdx:570]: I did spend time investigating this a few months ago, you can see the details here: https://github.com/pytorch/pytorch/issues/71274

the pytorch devs swear it's supposed to be faster, but I wasn't able to see any speed improvement. tldr: https://github.com/pytorch/pytorch/issues/71274#issuecomment-1047438622

I have a separate PR that expands the optimizers section: https://github.com/huggingface/transformers/pull/14708 but I was thinking to add it after the dust settles, but I suppose it might be better to put it into the "fodder" now?

Comment (sgugger) [reply docs/source/en/perf_train_gpu_one.mdx:570]: As long as it's not forgotten here I'm fine with bot solutions.

File: docs/source/en/_toctree.yml (hunk 1 above)
Comment (lvwerra) [reply docs/source/en/_toctree.yml:81]: Good catch!must have happened during rebase - I'll remove it.

File: docs/source/en/perf_train_gpu_many.mdx (hunk 4 above)
Comment (lvwerra) [reply docs/source/en/perf_train_gpu_many.mdx:72]: I added a warning sentence. Is that what you had in mind?

File: docs/source/en/perf_train_gpu_one.mdx (hunk 9 above)
Comment (lvwerra) [reply docs/source/en/perf_train_gpu_one.mdx]: Thanks, removed this for now.

File: docs/source/en/perf_train_gpu_many.mdx [hunk 12]
```diff
@@ -1,47 +1,157 @@
...
-4. Zero Redundancy Optimizer (ZeRO) - Also performs sharding of the tensors somewhat similar to TP, except the whole tensor gets reconstructed in time for a forward or backward computation, therefore the model doesn't need to be modified. It also supports various offloading techniques to compensate for limited GPU memory.
-5. Sharded DDP - is another name for the foundational ZeRO concept as used by various other implementations of ZeRO.
 
 
 ## Data Parallelism
 
-Most users with just 2 GPUs already enjoy the increased training speed up thanks to `DataParallel` (DP) and `DistributedDataParallel` (DDP) that are almost trivial to use. This is a built-in feature of Pytorch.
+Most users with just 2 GPUs already enjoy the increased training speed up thanks to `DataParallel` (DP) and `DistributedDataParallel` (DDP) that are almost trivial to use. This is a built-in feature of Pytorch. Note that in general it is adviced to use DDP as it is better maintained and works for all models while DP might fail for some models.
```
Comment (sgugger) [inline docs/source/en/perf_train_gpu_many.mdx:69]:
```suggestion
Most users with just 2 GPUs already enjoy the increased training speed up thanks to `DataParallel` (DP) and `DistributedDataParallel` (DDP) that are almost trivial to use. This is a built-in feature of Pytorch. Note that in general it is advised to use DDP as it is better maintained and works for all models while DP might fail for some models. [PyTorch documentation](https://pytorch.org/docs/master/generated/torch.nn.DataParallel.html) itself recommends the use of DDP.
```

File: docs/source/en/perf_train_gpu_many.mdx (hunk 4 above)
Comment (sgugger) [reply docs/source/en/perf_train_gpu_many.mdx:72]: Just added a suggestion on it, thanks!

**Comment by stas00** [comment]
@lvwerra, I rebased and solved the conflict introduced by https://github.com/huggingface/transformers/pull/16860 - fixing it in the new location.

**Comment by lvwerra** [comment]
@stas00 thank you!

@sgugger @LysandreJik would you like to take a look at the PR? I integrated the previous suggestion.

**Review by LysandreJik — APPROVED** [review]
Impressive refactor!

**Comment by lvwerra** [comment]
@sgugger are you happy to merge this?

File: docs/source/en/_toctree.yml [hunk 13]
```diff
@@ -83,6 +79,14 @@
...
   - local: add_new_pipeline
     title: "How to add a pipeline to 🤗 Transformers?"
+  - local: performance
+    title: Performance and scalability
+  - local: perf_train_gpu_one
+    title: Training on one GPU
+  - local: perf_train_gpu_many
+    title: Training on many GPUs
+  - local: perf_hardware
+    title: Custom hardware for training
```
Comment (sgugger) [inline docs/source/en/_toctree.yml:89]: This ends up being a bit too low in the toc for my taste. I think at least the perforamce page should be where it was before.

**Review by sgugger — APPROVED** [review]

**Comment by sgugger** [comment]
Yes, LGTM :-)

File: docs/source/en/_toctree.yml (hunk 13 above)
Comment (lvwerra) [reply docs/source/en/_toctree.yml]: Moved it back up :)
